crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1747,22 +1748,40 @@ impl<'db> TypeInferenceBuilder<'db> {
         iterable: &ast::Expr,
         target: &ast::ExprName,
         is_first: bool,
+        is_async: bool,
         definition: Definition<'db>,
     ) {
-        if !is_first {
-            let expression = self.index.expression(iterable);
-            let result = infer_expression_types(self.db, expression);
+        let expression = self.index.expression(iterable);
+        let result = infer_expression_types(self.db, expression);
+
+        // Two things are different if it's the first comprehension:
+        // (1) We must lookup the type of the expression in the outer scope
+        //     (symbols defined in this scope are not available for definitions
+        //     created by the first comprehension)
+        // (2) We must *not* call `self.extend()` on the result of the type inference,
+        //     or we'll doubly infer the type of relevant symbols
+        //     (any symbols referenced from the first comprehension will have already been
+        //     inferred when analysing outer scopes)
```

Comment:
```suggestion
        // Two things are different if it's the first comprehension:
        // (1) We must lookup the `ScopedExpressionId` of the iterable expression in the outer scope,
        //     because that's the scope we visit it in in the semantic index builder
        // (2) We must *not* call `self.extend()` on the result of the type inference,
        //     because `ScopedExpressionId` are only meaningful within their own scope, so
        //     we'd add types for random wrong expressions in the current scope
```